### PR TITLE
Reland: Framework wide color (#54415)

### PR DESCRIPTION
**Pull request by gaaclarke**
[This PR](https://github.com/flutter/engine/pull/54415) was reverted because it requires a manual roll into the framework.

issue: https://github.com/flutter/flutter/issues/127855
integration test: https://github.com/flutter/engine/pull/54415

This does the preliminary work for implementing wide gamut colors in the Flutter framework. Here are the following changes: 1) colors now specify a colorspace with which they are to be interpreted 1) colors now store their components as floats to accommodate bit depths more than 8

The storage of this Color class is weird with float/int storage but that is a temporary solution to support a smooth transition. Here is the plan for landing this:
1) Land this PR
1) Wait for it to roll into the Framework
1) Land https://github.com/flutter/flutter/pull/153938 which will make CupertinoDynamicColor implement Color
1) Land another engine PR that rips out the int storage: https://github.com/flutter/engine/pull/54714

Here are follow up PRs:
1) https://github.com/flutter/engine/pull/54473 - changes DlColor so the wide gamut colors are rendered
1) https://github.com/flutter/engine/pull/54567 - Hooks up these changes to take advantage of wide DlColor
1) https://github.com/flutter/flutter/pull/153319 - the integration test for the framework repo

There are some things that have been left as follow up PRs since they are technically breaking:
1) The math on `lerp` hasn't been updated to take advantage of the higher bit depth
1) `operator==` hasn't been updated to take advantage of the higher bit depth
1) `hashCode` hasn't been updated to take advantage of the higher bit depth
1) `alphaBlend` hasn't been updated to take advantage of the higher bit depth
1) `toString` hasn't been updated to take advantage of the higher bit depth

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by jonahwilliams** [comment]
I the problem was the customer tests? One of the dependencies in there is https://pub.dev/packages/flutter_color_models which has a bunch of Color implementations

**Comment by gaaclarke** [comment]
I'm going to try to preload a framework patch as best I can from looking at the log before landing this.  It is possible it's impossible to get 100% right before landing since some tests may be short circuited.

**Comment by gaaclarke** [comment]
> I the problem was the customer tests? One of the dependencies in there is https://pub.dev/packages/flutter_color_models which has a bunch of Color implementations

I haven't dug into it yet, just starting to look now.  I'll report back.

**Comment by gaaclarke** [comment]
If the problem is the change in the interface I think we can potentially preload the changes as long as missing `@override` isn't a CI breaking lint.

**Comment by gaaclarke** [comment]
Okay yea, all the errors seem to be coming from customer testing for [that one color package](https://pub.dev/packages/flutter_color_models).

**Comment by gaaclarke** [comment]
Everything else looked fine.  Here's the failed roll: https://github.com/flutter/flutter/pull/153985

**Comment by jonahwilliams** [comment]
You should be able to add the overrides and // ignore  any warnings.

**Comment by gaaclarke** [comment]
> You should be able to add the overrides and // ignore any warnings.

@jonahwilliams The customer tests are running against the latest published version?  I'll try posting a PR for the dev and see how responsive they are.

**Comment by jonahwilliams** [comment]
Its whatever is in the pubspec. If it comes to it, we can publish a fork and add a dependency override or something like. Please also email the owner of that test as FYI, I think they're generally fairly responsivle

**Comment by gaaclarke** [comment]
I've published a PR for that package: https://github.com/james-alex/color_models/pull/10. I haven't heard back from the package author yet.

**Comment by gaaclarke** [comment]
@jonahwilliams this should be good to land now, the test have been upgraded in https://github.com/flutter/tests/pull/404.  I was able to test color_model tests against master and this and they ran correctly locally, so I think we are good.  Still no update from the color_models maintainer.

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Review by matanlurey — APPROVED** [review]

File: lib/ui/lerp.dart
Comment (jonahwilliams) [inline lib/ui/lerp.dart:29]: I think this is what may be causing the breakages. Throw inf in here for a or b and some lerp cases now produce nans:

image

**Comment by feinstein** [comment]
@gaaclarke one thing is not clear to me, after reading some docs and issues.... will a Flutter app, running on iOS, automatically switch all it's colors to P3, or do we have to configure this manually for each Color we create?

**Comment by matanlurey** [comment]
> manually for each Color we create?

This.

It would be a breaking change to silently change your colors.

See https://docs.flutter.dev/release/breaking-changes/wide-gamut-framework#8-bit-unsigned-integer-constructors:

> Constructors like Color.fromARGB remain unchanged and have continued support. To take advantage of Display P3 colors, you must use the new Color.from constructor that takes normalized floating-point color components

**Comment by feinstein** [comment]
Can we get a global flag, like material 3 had in theme, to automatically use p3 for the entire app?

**Comment by jonahwilliams** [comment]
@feinstein You are using P3 for the entire app. P3 doesn't make existing colors more vibrant. it just adds some "extra" colors  around the edges of the gamut. If you want to use those colors you need to specify them. If you ignore the existing color space and pretend, for example, that all the material colors are in p3 space - then you get an entirely different set of colors than was intended by the designers.

**Comment by matanlurey** [comment]
I am locking this thread, `flutter/engine` is about to get archived, and a PR started in August 2024 is not the place to have this discussion. If you have a specific request ("I want a tool that ___" or "my colors don't look as expected in ___"), you can file a new issue at https://github.com/flutter/flutter (though I'd expect we would _not_ build a tool to do this, for the reason @jonahwilliams mentions already).

**Pull request merged**
